File: android/app/src/main/java/com/screenmeet/sdkdemo/recycler/ParticipantsAdapter.java
package com.screenmeet.sdkdemo.recycler;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.screenmeet.sdk.Identity;
import com.screenmeet.sdk.Participant;
import com.screenmeet.sdkdemo.R;
import com.screenmeet.sdkdemo.databinding.PartcipantLayoutBinding;

import org.webrtc.EglBase;
import org.webrtc.RendererCommon;
import org.webrtc.SurfaceViewRenderer;
import org.webrtc.VideoTrack;

import java.util.ArrayList;

public class ParticipantsAdapter extends RecyclerView.Adapter<ParticipantsAdapter.ViewHolder> {

    private final ArrayList<Participant> participants;
    private final EglBase eglBase;

    public ParticipantsAdapter(ArrayList<Participant> participants, EglBase eglBase) {
        this.participants = participants;
        this.eglBase = eglBase;
    }

    @Override
    public ParticipantsAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        LayoutInflater inflater = LayoutInflater.from(context);

        PartcipantLayoutBinding binding = PartcipantLayoutBinding.inflate(inflater);
        binding.surfaceViewRenderer.init(eglBase.getEglBaseContext(), null);
        binding.surfaceViewRenderer.setZOrderMediaOverlay(true);
        binding.surfaceViewRenderer.setScalingType(RendererCommon.ScalingType.SCALE_ASPECT_FIT);
        return new ViewHolder(binding.getRoot());
    }

    public void add(Participant participant) {
        participants.add(participant);
        notifyItemInserted(participants.size() - 1);
    }

    public void remove(Participant participant) {
        int index = participants.indexOf(participant);
        if(index != -1){
            participants.remove(index);
            notifyItemRemoved(index);
        }
    }

    public void update(Participant participant) {
        int index = participants.indexOf(participant);
        if(index != -1){
            participants.set(index, participant);
            notifyItemChanged(index);
        }
    }

    @Override
    public void onBindViewHolder(ParticipantsAdapter.ViewHolder holder, int position) {
        Participant participant = participants.get(position);

        holder.nameTv.setText(participant.getIdentity().getName());

        if(participant.getIdentity().getRole() == Identity.Role.HOST){
            holder.hostImage.setVisibility(View.VISIBLE);
        } else holder.hostImage.setVisibility(View.GONE);

        if(participant.getCallerState().getAudioEnabled()){
            holder.micButton.setImageResource(R.drawable.mic);
        } else holder.micButton.setImageResource(R.drawable.mic_off);

        boolean videoEnabled = participant.getCallerState().getVideoEnabled();
        boolean screenEnabled = participant.getCallerState().getScreenEnabled();
        if(videoEnabled || screenEnabled){
            if(videoEnabled) holder.cameraButton.setImageResource(R.drawable.videocam);
            else holder.cameraButton.setImageResource(R.drawable.screenshot);
        } else {
            holder.cameraButton.setImageResource(R.drawable.videocam_off);
        }

        VideoTrack videoTrack = participant.getVideoTrack();
        if (videoTrack != null) {
            videoTrack.setEnabled(true);
            videoTrack.addSink(holder.renderer);
        }
    }

    // Returns the total count of items in the list
    @Override
    public int getItemCount() {
        return participants.size();
    }


    public static class ViewHolder extends RecyclerView.ViewHolder {

        public TextView nameTv;
        public ImageView micButton;
        public ImageView cameraButton;
        public ImageView hostImage;
        public SurfaceViewRenderer renderer;

        public ViewHolder(View itemView) {
            super(itemView);

            nameTv = (TextView) itemView.findViewById(R.id.nameTv);
            micButton = (ImageView) itemView.findViewById(R.id.microButton);
            cameraButton = (ImageView) itemView.findViewById(R.id.cameraButton);
            hostImage = (ImageView) itemView.findViewById(R.id.hostImage);
            renderer = (SurfaceViewRenderer) itemView.findViewById(R.id.surfaceViewRenderer);
        }
    }
}

File: android/app/src/main/java/com/screenmeet/live/feature/call/ParticipantsAdapter.kt
package com.screenmeet.live.feature.call

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.core.content.ContextCompat
import androidx.core.view.isVisible
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.screenmeet.live.R
import com.screenmeet.live.databinding.LayoutParticipantBinding
import com.screenmeet.live.tools.DoubleTapListener
import com.screenmeet.sdk.VideoElement
import kotlinx.coroutines.CoroutineScope
import org.webrtc.EglBase
import org.webrtc.RendererCommon

typealias PinClick = (VideoElement) -> Unit

class ParticipantsAdapter(
    private val scope: CoroutineScope,
    private val eglBase: EglBase.Context,
    private val onClick: PinClick
) : ListAdapter<VideoElement, ParticipantsAdapter.ViewHolder>(UiVideoComparator()) {

    val itemSpacing = 30

    private var recyclerSize = 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val context = parent.context
        val inflater = LayoutInflater.from(context)
        recyclerSize = parent.measuredWidth
        val binding = LayoutParticipantBinding.inflate(inflater)
        binding.renderer.init(
            eglBase,
            object : RendererCommon.RendererEvents {
                override fun onFrameResolutionChanged(
                    videoWidth: Int,
                    videoHeight: Int,
                    rotation: Int
                ) {
                    binding.renderer.post {
                        binding.renderer.layoutParams = fitFrame(
                            videoWidth,
                            videoHeight,
                            binding.root.width,
                            binding.root.height
                        )
                    }
                }

                override fun onFirstFrameRendered() {
                }
            }
        )
        return ViewHolder(binding)
    }

    private fun fitFrame(
        videoWidth: Int,
        videoHeight: Int,
        containerWidth: Int,
        containerHeight: Int
    ): FrameLayout.LayoutParams {
        val videoAspect = videoWidth.toFloat() / videoHeight
        val containerAspect = containerWidth.toFloat() / containerHeight

        val newWidth: Int
        val newHeight: Int

        if (videoAspect > containerAspect) {
            newWidth = containerWidth
            newHeight = (containerWidth / videoAspect).toInt()
        } else {
            newHeight = containerHeight
            newWidth = (containerHeight * videoAspect).toInt()
        }

        val offsetX = (containerWidth - newWidth) / 2
        val offsetY = (containerHeight - newHeight) / 2

        return FrameLayout.LayoutParams(newWidth, newHeight).apply {
            setMargins(offsetX, offsetY, 0, 0)
        }
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val participant = getItem(position)
        holder.bind(scope, participant, recyclerSize - itemSpacing * 3, onClick)
    }

    override fun onViewRecycled(holder: ViewHolder) {
        holder.dispose()
    }

    fun dispose(recycler: RecyclerView) {
        currentList.forEachIndexed { i, _ ->
            val viewHolder = recycler.findViewHolderForAdapterPosition(i) as? ViewHolder
            viewHolder?.dispose()
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    class ViewHolder(val binding: LayoutParticipantBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun bind(scope: CoroutineScope, video: VideoElement, recyclerSize: Int, onClick: PinClick) {
            binding.apply {
                val layoutParams = RecyclerView.LayoutParams(
                    recyclerSize / 2,
                    recyclerSize / 2
                )
                itemView.layoutParams = layoutParams

                val videoTrack = video.track
                val hasTrack = videoTrack != null
                renderer.render(videoTrack)

                if (hasTrack) {
                    renderer.listenFramesStuck(scope) { stuck ->
                        frameStuckSpinner.isVisible = stuck
                    }
                } else {
                    frameStuckSpinner.isVisible = false
                }

                nameTv.text = video.userName
                logo.isVisible = !hasTrack
                renderer.isVisible = hasTrack
                root.setOnTouchListener(
                    DoubleTapListener(root.context) {
                        onClick(video)
                    }
                )

                pinButton.setOnClickListener {
                    onClick(video)
                }

                if (video.isAudioSharing) {
                    microButton.setImageResource(R.drawable.mic)
                    microButton.backgroundTintList = null
                    microButton.colorFilter = null
                } else {
                    val context = root.context
                    microButton.setImageResource(R.drawable.mic_off)
                    val color = ContextCompat.getColor(context, R.color.error_red)
                    microButton.setColorFilter(color)
                }
            }
        }

        fun dispose() {
            binding.renderer.clear()
        }
    }

    internal class UiVideoComparator : DiffUtil.ItemCallback<VideoElement>() {
        override fun areItemsTheSame(oldItem: VideoElement, newItem: VideoElement): Boolean =
            oldItem.id == newItem.id

        override fun areContentsTheSame(oldItem: VideoElement, newItem: VideoElement): Boolean =
            oldItem == newItem
    }
}
